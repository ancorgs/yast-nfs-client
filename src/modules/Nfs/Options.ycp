{

module "Nfs::Options";
textdomain "nfs";

/**
 * Checks the nfs options for /etc/fstab:
 * nonempty, comma separated list of foo,nofoo,bar=baz (see nfs(5))
 * @param options   options
 * @return          a translated string with error message, emtpy string if ok
 */
global define string validate (string options) {
    // To translators: error popup
    if (size (options) == 0)   return _("Empty option strings are not allowed.");
    if (options == "defaults") return "";


    list<string> option_list = splitstring (options, ",");

    //the options must be easy to sync with mount.c and nfsmount.c

    // these can be negated by "no"
    list<string> non_value   = ["bg", "fg", "soft", "hard", "intr", "posix", "cto", "ac", "acl",
    		    "lock", "tcp", "udp", "rdirplus",
    		    // these are common for all fs types
    		    "atime", "auto", "dev", "exec", "group", "owner",
    			    "suid", "user", "users"];
    // these cannot be negated
    // they are not nfs specific BTW
    list non_value1 = ["defaults", "async", "sync", "dirsync", "ro", "rw",
    		   "remount", "bind", "rbind", "_netdev", ];
    list with_value  = ["rsize", "wsize", "timeo", "retrans", "acregmin", "acregmax",
    		    "acdirmin", "acdirmin", "acdirmax", "actimeo", "retry", "namlen",
    		    "port", "proto", "clientaddr", "mountport", "mounthost",
    		    "mountprog", "mountvers", "nfsprog", "nfsvers", "vers", "minorversion",
                        "sec" ];
    integer i = 0;
    string current_option = "";

    // first fiter out non value options and its nooptions forms (see nfs(5))
    option_list = filter (string e, option_list, ``(!contains (non_value, e)));
    non_value   = maplist (string e, non_value, ``(sformat ("no%1", e)));
    option_list = filter (string e, option_list, ``(!contains (non_value, e)));
    option_list = filter (string e, option_list, ``(!contains (non_value1, e)));

    while (i < size (option_list))
    {
        string opt = option_list[i]:"";
        list<string> value = splitstring (opt, "=");
        string v0 = value[0]:"";
        string v1 = value[1]:"";
        // FIXME: this also triggers for "intr=bogus"
        // To translators: error popup
        if (!contains (with_value, v0)) return sformat (_("Unknown option: %1"), v0);
        // To translators: error popup
        if (size (value) != 2) return sformat (_("Invalid option: %1"), opt);
        // To translators: error popup
        if (v1 == "") return sformat (_("Empty value for option: %1"), v0);
        i = i + 1;
    }

    return "";
}

/**
FIXME: factor out get_nfs4(vfstype, options) (depending on n::o)!
 * @return is version >= 4.1 enabled
 */
global boolean get_nfs41(string options) {
    list<string> option_list = splitstring (options, ",");

    return contains(option_list, "minorversion=1");
}

/**
 * Add or remove minorversion=1 according to nfs41.
 * FIXME vfstype=nfs4 is deprecated in favor of nfsvers=4 (aka vers=4)
 */
global define string set_nfs41(string options, boolean nfs41) {
    // don't mutate the string unnecessarily
    if (get_nfs41(options) == nfs41) {
        return options;
    }

    const string OPT   = "minorversion=1";
    const string CLEAN = "minorversion=0";

    if (options == "defaults") { options = ""; }
    list<string> option_list = splitstring(options, ",");
    option_list = filter(string opt, option_list, ``( opt != OPT ));
    option_list = filter(string opt, option_list, ``( opt != CLEAN ));

    if (nfs41) {
        option_list = add(option_list, OPT);
    }

    options = mergestring(option_list, ",");
    if (options == "") { options = "defaults"; }

    return options;
}

/*EOF*/
}
