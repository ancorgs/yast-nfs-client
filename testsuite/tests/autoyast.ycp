{
    include "testsuite.ycp";
    import "Nfs";
    import "Assert";

    DUMP("Nfs::Import");
    // ---------
    DUMP("- basic, SLE11-SP2");
    map entry1 = $[
                   "server_path": "data.example.com:/mirror",
                   "mount_point": "/mirror",
                   "nfs_options": "defaults"
                   ];
    
    Nfs::Import([ entry1 ]);
    Assert::Equal(1, size(Nfs::nfs_entries));
    Assert::Equal("data.example.com:/mirror", Nfs::nfs_entries[0, "spec"]:"");

    DUMP("-- and Export");
    list e = Nfs::Export();
    Assert::Equal(1, size(e));
    Assert::Equal("data.example.com:/mirror", e[1, "server_path"]:"");
    Assert::Equal("/mirror",                  e[1, "mount_point"]:"");
    Assert::Equal("defaults",                 e[1, "nfs_options"]:"");

    // ---------
    DUMP("- empty");
    Nfs::Import([]);
    Assert::Equal(0, size(Nfs::nfs_entries));

    // ---------
    DUMP("- invalid, missing basic data");
    map entry_invalid = $[
                   "server_path": "data.example.com:/mirror",
                   ];
    
    Nfs::Import([ entry_invalid ]);
    Assert::Equal(0, size(Nfs::nfs_entries));
}
