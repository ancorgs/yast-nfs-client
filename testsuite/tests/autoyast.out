Read	.target.tmpdir "/tmp"
Dump	Nfs::Import
Dump	- basic, SLE11-SP2
Read	.sysconfig.nfs.NFS4_SUPPORT "no"
Read	.etc.idmapd_conf "Domain" "localdomain"
Return	true
Dump	-- and Export
Dump	- empty
Read	.sysconfig.nfs.NFS4_SUPPORT "no"
Read	.etc.idmapd_conf "Domain" "localdomain"
Return	true
Dump	- invalid, missing basic data
Log	Missing at Import: 'mount_point'.
Log	Missing at Import: 'nfs_options'.
Dump	- basic, SLE11-SP3
Dump	-- and Export
